File: lib/mac/axapi_node.swift
import ApplicationServices

public class AXAPINode {
    let ax_element: AXUIElement

    init(_ ax_element: AXUIElement) {
      self.ax_element = ax_element
    }

    public var role: String? {
      return getStringAttribute(kAXRoleAttribute)
    }

    public var title: String? {
      return getStringAttribute(kAXTitleAttribute)
    }

    func getAttributeNames() -> [String] {
        var attributes: CFArray?
        let result = AXUIElementCopyAttributeNames(ax_element, &attributes)
        guard result == .success else {
            print("Couldn't get attributes")
            return []
        }

        return attributes! as! [String]
    }

    func getStringAttribute(_ attribute: String) -> String? {
        var value: CFTypeRef?
        let result = AXUIElementCopyAttributeValue(ax_element,
                                                   attribute as CFString,
                                                   &value)
        guard result == .success else {
            print("Couldn't get \(attribute)")
            return nil
        }
        if let string_value = value as? String {
            return string_value
        }
        print("\(attribute) could not be converted to String")
        return nil
    }
}

public func createAXAPINodeForApplication(_ pid: pid_t) -> AXAPINode {
  let pid_ax_element = AXUIElementCreateApplication(pid)
  return AXAPINode(pid_ax_element)
}

File: lib/mac/axapi_node.cc
#include "include/axaccess/mac/axapi_node.h"

#include <string>
#include <vector>

namespace mac_inspect {

AXAPINode::AXAPINode(AXUIElementRef ax_element) : ax_element_(ax_element) {}

std::string AXAPINode::GetRole() {
  return "";
}

std::string AXAPINode::GetTitle() {
  return "";
}

std::vector<std::string> AXAPINode::GetAttributeNames() {
  std::vector<std::string> result;
  return result;
}

std::string AXAPINode::GetStringAttributeValue(std::string& attribute_name) {
  return "";
}

AXAPINode AXAPINode::createForPID(pid_t pid) {
  AXUIElementRef axuielement = AXUIElementCreateApplication(pid);
  return AXAPINode(axuielement);
}

}  // namespace mac_inspect